File: JGSourceBaseDemo/JGSourceBaseDemo/Classes/Base/JGSDViewController.m
//
//  JGSDViewController.m
//  JGSourceBaseDemo
//
//  Created by Mei JiGao on 2026/6/17.
//  Copyright © 2026 ByMountains. All rights reserved.
//

#import "JGSDViewController.h"
#import "JGSourceBaseDemo-Swift.h"

@interface JGSDViewController ()

/// 测试入口列表，默认根据sections、rows数据情况控制显示、隐藏
@property (nonatomic, strong) UITableView *tableView;
/// 页面内容较多时，使用该容器展示页面元素，默认隐藏
@property (nonatomic, strong) UIScrollView *scrollView;
/// 页面底部日志展示窗口，顶部接tableView、scrollView底部
@property (nonatomic, strong) UITextView *logTextView;

// 列表数据
@property (nonatomic, copy) NSArray<JGSDTableSectionData *> *sections;
@property (nonatomic, copy) NSArray<JGSDTableCellData *> *rows;

@end

@implementation JGSDViewController

- (void)setTitle:(NSString *)title {
    [super setTitle:title];
}

- (void)setSubtitle:(NSString *)subtitle {
    _subtitle = subtitle;
}

- (UITableView *)tableView {
    
    if (_tableView) {
        return _tableView;
    }
    
    UITableView *v = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStyleGrouped];
    v.hidden = YES;
    v.backgroundColor = self.view.backgroundColor;
    v.alwaysBounceVertical = YES;
    v.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentAlways;
    
    v.sectionHeaderHeight = UITableViewAutomaticDimension;
    v.rowHeight = UITableViewAutomaticDimension;
    v.sectionFooterHeight = CGFLOAT_MIN;
    
    v.dataSource = self;
    v.delegate = self;
    [v registerClass:UITableViewCell.class forCellReuseIdentifier:@"UITableViewCell"];
    
    return _tableView = v;
}

- (UIScrollView *)scrollView {
    
    if (_scrollView) {
        return _scrollView;
    }
    
    UIScrollView *v = [[UIScrollView alloc] init];
    v.hidden = YES;
    v.backgroundColor = self.view.backgroundColor;
    v.alwaysBounceVertical = YES;
        
    // 撑开宽度
    UIView *line = [[UIView alloc] init];
    [v addSubview:line];
    [line mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.top.right.mas_equalTo(v);
        make.height.mas_equalTo(0);
        make.width.mas_equalTo(v);
    }];
    
    return _scrollView = v;
}

- (UITextView *)logTextView {
    
    if (_logTextView) {
        return _logTextView;
    }
    
    UITextView *v = [[UITextView alloc] init];
    v.backgroundColor = [UIColor whiteColor];
    v.editable = NO;
    v.textColor = [UIColor darkGrayColor];
    v.font = [UIFont systemFontOfSize:16];
    v.contentInset = UIEdgeInsetsMake(5, 5, 5, 5);
    
    v.attributedText = [[NSAttributedString alloc] initWithString:@"调试日志输出区域，内容可复制、不可编辑" attributes:@{NSForegroundColorAttributeName: [UIColor lightGrayColor]}];
    
    return _logTextView = v;
}

// MARK: - Data
- (void)setSections:(NSArray<JGSDTableSectionData *> *)sections {
    _sections = sections ?: @[];
}

- (void)setRows:(NSArray<JGSDTableCellData *> *)rows {
    _rows = rows ?: @[];
}

- (void)loadData {
    
}

- (void)setupData:(NSArray<JGSDTableSectionData *> *)sections rows:(NSArray<JGSDTableCellData *> *)rows {
    if (sections) {
        self.sections = sections;
    }
    if (rows) {
        self.rows = rows;
    }
    
    NSArray<JGSDTableCellData *> *secRows = [self.sections flatMap:^NSArray * _Nonnull(JGSDTableSectionData * _Nonnull obj) {
        return obj.rows;
    }];
    self.tableView.hidden = secRows.count + self.rows.count == 0;
}

// MARK: - Controller
- (void)dealloc {
    JGSLog(@"<%@: %p>", NSStringFromClass([self class]), self);
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Do any additional setup after loading the view.
    self.edgesForExtendedLayout = UIRectEdgeNone;
    self.view.backgroundColor = [UIColor colorWithWhite:0.96 alpha: 1.0];
    
    // title
    self.title = self.title ?: [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleName"] ?: NSStringFromClass([self class]);
    NSString *version = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"];
    NSString *build = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleVersion"];
    self.subtitle = [NSString stringWithFormat:@"%@ (%@)", version, build];
    
    [self loadData];
    [self setupViews];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
//#pragma clang diagnostic push
//#pragma clang diagnostic ignored "-Wdeprecated-declarations"
//    [JGSLogFunction enableLog:!JGSLogFunction.isLogEnabled];
//#pragma clang diagnostic pop
//    [JGSLogger enableLogWithMode:!JGSLogger.enableDebug level:JGSLogger.level useNSLog:JGSLogger.useNSLog];
//    [JGSLogger enableLogWithMode:!JGSLogger.enableDebug level:JGSLogger.level useNSLog:JGSLogger.useNSLog lengthLimit:JGSLogger.lengthLimit truncating:JGSLogger.truncating];
    
    JGSLog(@"%.02f", 0.926);
}

// MARK: - UI
- (void)setupViews {
    
    [self.view addSubview:self.scrollView];
    [self.scrollView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.left.right.equalTo(self.view);
    }];
    
    [self.view addSubview:self.tableView];
    [self.tableView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.left.right.equalTo(self.view);
    }];
    
    [self.view addSubview:self.logTextView];
    [self.logTextView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.tableView.mas_bottom);
        make.top.equalTo(self.scrollView.mas_bottom);
        make.left.bottom.right.equalTo(self.view);
        make.height.equalTo(self.view).multipliedBy(0.25);
    }];
}

// MARK: - Console
- (void)showConsoleLog:(NSString *)format, ... {
    
    va_list varList;
    va_start(varList, format);
    NSString *message = [[NSString alloc] initWithFormat:format arguments:varList];
    va_end(varList);
    
    self.logTextView.text = [self.logTextView.text stringByAppendingFormat:@"\n%@", message];
    [self.logTextView scrollRangeToVisible:NSMakeRange(self.logTextView.text.length - 1, 1)];
}

#pragma mark - UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.sections.count > 0 ? self.sections.count : 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.sections.count > 0 ? self.sections[section].rows.count : self.rows.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"UITableViewCell" forIndexPath:indexPath];
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    
    JGSDTableCellData *rowData = self.sections.count > 0 ? self.sections[indexPath.section].rows[indexPath.row] : self.rows[indexPath.row];
    cell.textLabel.text = rowData.title;
    
    return cell;
}

#pragma mark - UITableViewDelegate
- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    if (self.sections.count > 0) {
        return self.sections[section].title.length == 0 ? CGFLOAT_MIN : UITableViewAutomaticDimension;
    }
    return CGFLOAT_MIN;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    if (self.sections.count > 0) {
        return self.sections[section].title;
    }
    return nil;
}

- (void)tableView:(UITableView *)tableView willDisplayHeaderView:(UIView *)view forSection:(NSInteger)section {
    if ([@"26.0" compare:UIDevice.currentDevice.systemVersion] == NSOrderedDescending) {
        return;
    }
    
    // 设置原始文本，覆盖系统的大写文本
    [(UITableViewHeaderFooterView *)view textLabel].text = [self tableView:tableView titleForHeaderInSection:section];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    JGSDShowConsoleLog(self, @"indexPath: %@", indexPath)
    JGSDTableCellData *rowData = self.sections.count > 0 ? self.sections[indexPath.section].rows[indexPath.row] : self.rows[indexPath.row];
    
    // action 响应
    if (rowData.action) {
        rowData.action(indexPath);
    }
    
    // target-selector 响应
    if (rowData.target && rowData.selector && [rowData.target respondsToSelector:rowData.selector]) {
        
        // OC多参数实现，避免警告
        //IMP imp = class_getMethodImplementation(rowData.target.class, rowData.selector); // 需要 target: NSObject
        IMP imp = [rowData.target methodForSelector:rowData.selector];
        // (target, selector, args...)
        id (*func)(id, SEL, NSIndexPath *) = (void *)imp;
        func(rowData.target, rowData.selector, indexPath);
    }
}

@end

File: JGSourceBaseDemo/JGSourceBaseDemo/Classes/Base/JGSDViewController.swift
//
//  JGSDViewController.swift
//  JGSourceBaseDemo
//
//  Created by Mei JiGao on 2026/6/16.
//  Copyright © 2026 ByMountains. All rights reserved.
//

import UIKit
import JGSourceBase
import SnapKit

/// 定义一个不暴露的泛型基类
/// 由于 Objective-C 无法理解 Swift 的泛型，这个类及其子类对 Objective-C 不可见的
class JGSDSwiftViewController<T>: UIViewController {
    deinit {
        JGSLog("<\(type(of: self)): \(Unmanaged.passUnretained(self).toOpaque())> dealloc")
        NotificationCenter.default.removeObserver(self)
    }
}
class JGSDViewController: JGSDSwiftViewController<UIViewController>, JGDispatchQueueOnceUUID {
    
    override var title: String? {
        set {
            super.title = newValue
        }
        get {
            super.title
        }
    }
    /// 副标题，用于设置多行标题
    var subtitle: String?
    
    // 列表数据
    private(set) var sections: [(title: String, rows: [JGSDTableCellData])] = []
    private(set) var rows: [JGSDTableCellData] = []
    
    /// 子类重写该方法调用 setupData (sections: rows: ) 设置页面数据
    /// 不需要调用super
    func loadData() {
        
    }
    
    /// final 禁止子类重写，仅允许调用
    /// sections、rows 必须有一个不为空
    final func setupData(sections: [(title: String, rows: [JGSDTableCellData])]? = nil, rows: [JGSDTableCellData]? = nil) {
        if let sections = sections {
            self.sections = sections
        }
        if let rows = rows {
            self.rows = rows
        }
        
        let secRows = self.sections.flatMap({ (title: String, rows: [JGSDTableCellData]) in
            rows
        }) as? [JGSDTableCellData] ?? []
        tableView.isHidden = secRows.count + self.rows.count == 0
    }
    
    /// 测试入口列表，默认根据sections、rows数据情况控制显示、隐藏
    private(set) lazy var tableView = {
        let v = UITableView(frame: .zero, style: .insetGrouped)
        v.isHidden = true
        v.backgroundColor = view.backgroundColor
        v.alwaysBounceVertical = true
        v.contentInsetAdjustmentBehavior = .never
        
        v.sectionHeaderHeight = UITableView.automaticDimension
        v.rowHeight = UITableView.automaticDimension
        v.sectionFooterHeight = CGFLOAT_MIN
        
        v.dataSource = self
        v.delegate = self
        v.register(UITableViewCell.self, forCellReuseIdentifier: "UITableViewCell")
        return v
    }()
    /// 页面内容较多时，使用该容器展示页面元素，默认隐藏
    private(set) lazy var scrollView = {
        let v = UIScrollView()
        v.isHidden = true
        v.backgroundColor = view.backgroundColor
        v.alwaysBounceVertical = true
        
        // 撑开宽度
        let line = UIView()
        v.addSubview(line)
        line.snp.makeConstraints { make in
            make.top.left.right.equalToSuperview()
            make.height.equalTo(0)
            make.width.equalToSuperview()
        }
        
        return v
    }()
    /// 页面底部日志展示窗口，顶部接tableView、scrollView底部
    private(set) lazy var logTextView = {
        let v = UITextView()
        v.backgroundColor = .white
        v.isEditable = false
        v.textColor = .darkGray
        v.font = UIFont.systemFont(ofSize: 16)
        v.contentInset = UIEdgeInsets(top: 5, left: 5, bottom: 5, right: 5)
        
        v.attributedText = NSAttributedString(string: "调试日志输出区域，内容可复制、不可编辑", attributes: [.foregroundColor: UIColor.lightGray])
        
        return v
    }()
    
    // MARK: - Controller
    deinit {
        DispatchQueue.jg_clearExecutedIdentifier(self.jg_uuid)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Do any additional setup after loading the view.
        edgesForExtendedLayout = []
        view.backgroundColor = UIColor(white: 0.96, alpha: 1.0)
        
        // title
        title = title ?? Bundle.main.object(forInfoDictionaryKey: "CFBundleName") as? String ?? "\(type(of: self))"
        if let version = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString"),
           let build = Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") {
            subtitle = "\(version) \(build)"
        }
        
        loadData()
        setupViews()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        showConsoleLog("<\(type(of: self)): \(Unmanaged.passUnretained(self).toOpaque())> viewWillAppear at:", Date())
        
        JGSLogFunction.enableLog(!JGSLogFunction.isLogEnabled)
        
        navigationController?.navigationBar.tintColor = .white
        navigationController?.navigationBar.barTintColor = .blue
        if #available(iOS 15.0, *) {
            // standardAppearance
            let standardAppearance = navigationController?.navigationBar.standardAppearance ?? UINavigationBarAppearance()
            standardAppearance.backgroundColor = navigationController?.navigationBar.barTintColor ?? UIColor(white: 0.8, alpha: 0.82)
            standardAppearance.titleTextAttributes = navigationController?.navigationBar.titleTextAttributes ?? JGSDTitleTextAttributes
            standardAppearance.backgroundEffect = nil
            standardAppearance.shadowColor = .clear
            standardAppearance.shadowImage = navigationController?.navigationBar.shadowImage
            
            navigationController?.navigationBar.standardAppearance = standardAppearance
            
            // scrollEdgeAppearance
            let scrollEdgeAppearance = navigationController?.navigationBar.scrollEdgeAppearance ?? UINavigationBarAppearance()
            scrollEdgeAppearance.backgroundColor = navigationController?.navigationBar.barTintColor ?? UIColor(white: 0.8, alpha: 0.82)
            scrollEdgeAppearance.titleTextAttributes = navigationController?.navigationBar.titleTextAttributes ?? JGSDTitleTextAttributes
            scrollEdgeAppearance.backgroundEffect = nil
            scrollEdgeAppearance.shadowColor = .clear
            scrollEdgeAppearance.shadowImage = navigationController?.navigationBar.shadowImage
            
            navigationController?.navigationBar.scrollEdgeAppearance = scrollEdgeAppearance
        }
        
        DispatchQueue.jg_once(jg_uuid) { [weak self] in
            JGSLog("jg_once:", self?.jg_uuid)
        }
        
        // 刷新数据
        tableView.reloadData()
    }
    
    // MARK: - UI
    func setupViews() {
        
        view.addSubview(scrollView)
        scrollView.snp.makeConstraints { make in
            make.top.left.right.equalToSuperview()
        }
        
        view.addSubview(tableView)
        tableView.snp.makeConstraints { make in
            make.top.left.right.equalToSuperview()
        }
        
        view.addSubview(logTextView)
        logTextView.snp.makeConstraints { make in
            make.top.equalTo(tableView.snp.bottom)
            make.top.equalTo(scrollView.snp.bottom)
            make.left.bottom.right.equalToSuperview()
            make.height.equalToSuperview().multipliedBy(0.25)
        }
    }

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}

// MARK: - Console
func JGSDShowConsoleLog(_ vcT: JGSDViewController? = nil, format: String? = nil, _ args: Any?..., file: String = #file, funcName: String = #function, lineNum : Int = #line) {
    guard let vcT = vcT else {
        JGSLog(format: format, args, file: file, funcName: funcName, lineNum: lineNum) // Xcode控制台日志
        return
    }
    vcT.showConsoleLog(format: format, args, file: file, funcName: funcName, lineNum: lineNum)
}

extension JGSDViewController {
    
    func showConsoleLog(format: String? = nil, _ args: Any?..., file: String = #file, funcName: String = #function, lineNum : Int = #line) {
        let message = JGSLogMessage(format: format, args)
        JGSLog(message, file: file, funcName: funcName, lineNum: lineNum) // Xcode控制台日志
        
        DispatchQueue.main.async { [weak self] in
            guard let `self` = self else { return }
            // 执行输出日志方法所在文件、方法、行号
            // 对文件路径进行处理，获取文件名
            let file = URL(fileURLWithPath: file).lastPathComponent
            let fileFuncLine = "[\(file) \(funcName)] Line: \(lineNum)"
            
            // 页面日志展示框日志
            let log = "\(fileFuncLine) \(message)"
            if var text = logTextView.text {
                text.append("\n\(log)")
                logTextView.text = text
            } else {
                logTextView.text = log
            }
            
            // 滚动到最新日志
            logTextView.scrollRangeToVisible(NSMakeRange(logTextView.text.count - 1, 1))
        }
    }
}

// MARK: - UITableViewDataSource & UITableViewDelegate
extension JGSDViewController: UITableViewDataSource, UITableViewDelegate {
    
    // MARK: - UITableViewDataSource
    func numberOfSections(in tableView: UITableView) -> Int {
        return sections.count > 0 ? sections.count : 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return sections.count > 0 ? sections[section].rows.count : rows.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "UITableViewCell", for: indexPath)
        cell.accessoryType = .disclosureIndicator
        
        let rowData = {
            sections.count > 0 ? sections[indexPath.section].rows[indexPath.row] : rows[indexPath.row]
        }()
        cell.textLabel?.text = rowData.title
        
        return cell
    }
    
    // MARK: - UITableViewDelegate
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        if sections.count > 0 {
            return sections[section].title.isEmpty ? CGFLOAT_MIN : UITableView.automaticDimension
        }
        return CGFLOAT_MIN
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        if sections.count > 0 {
            return sections[section].title
        }
        return nil
    }
    
    func tableView(_ tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
        guard let headerView = view as? UITableViewHeaderFooterView else { return }
        guard UIDevice.current.systemVersion < "26.0" else {
            return
        }
        
        // 设置原始文本，覆盖系统的大写文本
        headerView.textLabel?.text = self.tableView(tableView, titleForHeaderInSection: section)
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        let rowData = {
            sections.count > 0 ? sections[indexPath.section].rows[indexPath.row] : rows[indexPath.row]
        }()
        
        // action 响应
        rowData.action?(indexPath)
        
        // target-selector 响应
        if let sel = rowData.selector {
            let target = sel.target
            let selector = sel.selector
            
            // 1. 获取 IMP (方法实现指针)
            // 注意：target 需为 AnyObject 并遵循 NSObjectProtocol，或者直接使用 NSObject 子类
            guard let imp = target.method(for: selector) else { return }
            // guard let imp = class_getMethodImplementation(object_getClass(target), action) else { return }
            // 2. 定义对应的 C 函数类型签名，各参数为 @objc 或者继承自 NSObject，注意返回值类型必须必须与实际匹配
            typealias ObjCMethodFunc = @convention(c) (AnyObject, Selector, IndexPath) -> Void
            // 3. 将 IMP 指针强制转换为可调用的函数类型
            let function = unsafeBitCast(imp, to: ObjCMethodFunc.self)
            // 4. 调用该函数
            function(target, selector, indexPath)
            
            // 此方案不需要验证参数为 @objc 或者继承自 NSObject
            // if target.responds(to: action) {
            //    _ = target.perform(action, with: indexPath)
            // }
        }
    }
}
